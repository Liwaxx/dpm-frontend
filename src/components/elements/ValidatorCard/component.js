import React from 'react';
import ButtonBase from '@material-ui/core/ButtonBase';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Modal from '@material-ui/core/Modal';
import Radio from '@material-ui/core/Radio';
import TextField from '@material-ui/core/TextField';
import moment from 'moment';
import PropTypes from 'prop-types';
import Text from '../Text';
import { IMAGES } from '../../../configs';

export default class Component extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      open: false,
      status: 'waiting',
    };

    moment.locale('id');
  }

  handleChange = prop => event => {
    this.setState({ [prop]: event.target.value });
  };

  handleClose = () => {
    this.setState({ open: false });
  };

  handleOpen = () => {
    this.setState({ open: true });
  };

  _renderModal() {
    const { classes, data } = this.props;
    const {
      accessor,
      description = [],
      document = [],
      nama_project,
      projectOwner,
      unit,
    } = data || {};
    const statusOptions = [
      { label: 'Waiting for Approval', value: 'waiting' },
      { label: 'Rejected', value: 'rejected' },
      { label: 'Approved', value: 'approved' },
    ];

    return (
      <Modal
        aria-labelledby="simple-modal-title"
        aria-describedby="simple-modal-description"
        open={this.state.open}
        onClose={this.handleClose}
      >
        <div className={classes.modalContainer}>
          <div className={classes.left}>
            <Text className={classes.modalTitle}>{nama_project}</Text>
            {this._renderModalLeftDescription(description)}
          </div>
          <div className={classes.right}>
            <ButtonBase
              disableRipple
              className={classes.exit}
              onClick={this.handleClose}
            >
              <img src={IMAGES.CLOSE} className={classes.close} />
            </ButtonBase>
            {this._renderModalRightText('P.O.', projectOwner)}
            {this._renderModalRightText('Unit', unit)}
            {this._renderModalRightDropdown('Status', statusOptions)}
            {this._renderModalRightRadio('Assesor', accessor, accessor)}
            {this._renderModalRightDocument('Document', document)}
            {this._renderModalRightDate('Attempt', document)}
          </div>
        </div>
      </Modal>
    );
  }

  _renderModalLeftDescription(data) {
    return (
      data.map((item, idx) => <Text key={idx}>{`- ${item.message}`}</Text>)
    );
  }

  _renderModalRightDate(name, options) {
    const { classes } = this.props;

    return (
      <div className={classes.modalRightContent}>
        <Text className={classes.modalRightName}>{name}</Text>
        {options.map((item, idx) => (
          <Text key={idx}>{`${idx + 1}. ${moment(item.attempt).format('DD-MM-YYYY')}`}</Text>
        ))}
      </div>
    );
  }

  _renderModalRightDocument(name, doc) {
    const { classes } = this.props;

    return (
      <div className={classes.modalRightContent}>
        <Text className={classes.modalRightName}>{name}</Text>
        <div className={classes.doc}>
          {doc.map((item, idx) => (
            <React.Fragment key={idx}>
              <ButtonBase href={item.url} target="blank">
                <div className={classes.docRed} />
              </ButtonBase>
            </React.Fragment>
          ))}
        </div>
      </div>
    );
  }

  _renderModalRightDropdown(name, options) {
    const { classes } = this.props;
    const { status } = this.state;
    const InputProps = {
      disableUnderline: true,
    };
    const SelectProps = {
      className: classes.select,
      native: true,
    };

    return (
      <div className={classes.modalRightContent}>
        <Text className={classes.modalRightName}>{name}</Text>
        <TextField
          select
          className={classes.dropdown}
          InputProps={InputProps}
          onChange={this.handleChange('status')}
          SelectProps={SelectProps}
          value={status}
        >
          {options.map(item => (
            <option
              className={classes.menu}
              key={item.value}
              value={item.value}
            >
              {item.label}
            </option>
          ))}
        </TextField>
      </div>
    );
  }

  _renderModalRightRadio(name, value, label) {
    const { classes } = this.props;
    const control = (<Radio
      checked={true}
      color="default"
      className={classes.radioButton}
    />);

    return (
      <div className={classes.modalRightContent}>
        <Text className={classes.modalRightName}>{name}</Text>
        <FormControlLabel
          value={value}
          control={control}
          label={label}
        />
      </div>
    );
  }

  _renderModalRightText(name, value) {
    const { classes } = this.props;

    return (
      <div className={classes.modalRightContent}>
        <Text className={classes.modalRightName}>{name}</Text>
        <Text>{value}</Text>
      </div>
    );
  }

  render() {
    const { classes, data } = this.props;
    const {
      description = [{}],
      modified_at,
      nama_project,
      projectOwner,
    } = data || {};

    return (
      <React.Fragment>
        {this._renderModal()}
        <div className={classes.main}>
          <ButtonBase onClick={this.handleOpen}>
            <div className={classes.card}>
              <div className={classes.heading}>
                <Text className={classes.title}>{nama_project}</Text>
                <Text className={classes.subTitle}>{projectOwner}</Text>
              </div>
              <Text className={classes.content}>{description[0].message || ''}</Text>
              <Text className={classes.subTitle}>{moment(modified_at).format('HH:mm')}</Text>
            </div>
          </ButtonBase>
        </div>
      </React.Fragment>
    );
  }
}

Component.propTypes = {
  classes: PropTypes.object,
  data: PropTypes.object,
};