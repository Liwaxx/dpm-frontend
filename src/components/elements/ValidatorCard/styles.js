import { VERTICAL_ALIGN } from '../../../styles';

const styles = {
  card: {
    backgroundColor: '#FFFFFF',
    borderRadius: '3px',
    display: 'flex',
    flexDirection: 'column',
    padding: '16px',
  },
  content: {
    color: '#B1B1B3',
    textAlign: 'left',
  },
  close: {
    height: '16px',
    width: '16px',
  },
  doc: {
    display: 'flex',
  },
  docRed: {
    backgroundColor: '#E2574C',
    height: '31px',
    marginRight: '8px',
    width: '27px',
  },
  dropdown: {
    border: '1px solid #AEB9CB',
    borderRadius: '2px',
  },
  exit: {
    height: '24px',
    justifyContent: 'flex-end',
    width: '100%',
  },
  heading: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'space-between',
  },
  left: {
    flex: '1 1 80%',
    overflowY: 'auto',
    padding: '16px',
  },
  main: {
    margin: '8px 0',
  },
  menu: {
    backgroundColor: '#ffffff',
    fontSize: '14px',
    padding: '0 8px',
  },
  modalContainer: {
    ...VERTICAL_ALIGN,
    backgroundColor: '#ffffff',
    display: 'flex',
    height: '80vh',
    margin: 'auto',
    width: '80%',
  },
  modalRightContent: {
    marginBottom: '16px',
  },
  modalRightName: {
    fontSize: '16px',
    fontWeight: 700,
  },
  modalTitle: {
    fontSize: '24px',
    marginBottom: '16px',
  },
  radioButton: {
    color: '#D8D8D8',
  },
  right: {
    borderLeft: '2px solid #DEE3ED',
    flex: '1 1 20%',
    padding: '16px',
  },
  select: {
    fontSize: '14px',
    padding: '0 8px',
  },
  subTitle: {
    fontSize: '14px',
    textAlign: 'right',
    color: '#8D98AA'
  },
  title: {
    fontSize: '20px',
    textAlign: 'left',
  },
};

export default styles;
