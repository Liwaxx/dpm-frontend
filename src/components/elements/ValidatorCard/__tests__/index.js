import React from 'react';
import renderer from 'react-test-renderer';
import ValidatorCard from '../index';

jest.mock('../../../elements/Text');

describe('ValidatorCard', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<ValidatorCard classes={{}} data={{}} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});