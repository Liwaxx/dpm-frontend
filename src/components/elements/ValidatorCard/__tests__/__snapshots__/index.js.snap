// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatorCard renders correctly 1`] = `
<div
  className="Component-main-10"
>
  <button
    className="MuiButtonBase-root-23"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <div
      className="Component-card-1"
    >
      <div
        className="Component-heading-8"
      />
    </div>
    <span
      className="MuiTouchRipple-root-26"
    />
  </button>
</div>
`;
